ocaml/xapi/pool_recommendations.ml

```diff
@@ -0,0 +1,58 @@
+(*
+ * Copyright (c) Cloud Software Group, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "pool_recommendations" end)
+
+open D
+module StringMap = Map.Make (String)
+
+let rec_map = ref StringMap.empty
+
+let finally f g =
+  try
+    let result = f () in
+    g () ; result
+  with e -> g () ; raise e
+
+let process_line data map =
+  match Xcp_service.Config_file.parse_line data with
+  | Some (key, v) ->
+      debug "Parsing data, key: %s, value: %s" key v ;
+      StringMap.add key v map
+  | None ->
+      map
+
+let parse filename =
+  debug "Parsing pool recommendations file: %s" filename ;
+  let ic = open_in filename in
```

Comment:
This code reads a file line by line. It clearly exists already - reuse that code.

Reply:
I saw that there is `parse` in `xcp_service`, but it has a different process function `process_line`. How can I reuse `parse` with different process logic?
```
  let parse filename spec =
    (* Remove the unnecessary doc parameter *)
    let ic = open_in filename in
    finally
      (fun () ->
        try
          while true do
            let line = input_line ic in
            process_line line spec
          done
        with End_of_file -> ()
      )
      (fun () -> close_in ic)
```

Reply:
Look at `uniext.ml`  and functions like `read_lines`, lines_fold`.

Reply:
Great function! Let me try it. Thank you!